### PR TITLE
4341 spot instances

### PR DESCRIPTION
**Pull request by gto389**
the deploy script is now capable for spot instance requests. This update also includes the cluster tagging as well.

### CODE REVIEW

File: src/encoded/commands/deploy.py [hunk 1]
```diff
@@ -117,26 +293,48 @@ def run(wale_s3_prefix, image_id, instance_type, elasticsearch, cluster_size, cl
         user_data = user_data % {
             'CLUSTER_NAME': cluster_name,
         }
-        security_groups = ['elasticsearch-https']
+        security_groups = ['ssh-http-https']
         iam_role = 'elasticsearch-instance'
         count = int(cluster_size)
+    
+    if not check_price == False :
```
Comment (hitz) [inline src/encoded/commands/deploy.py]: this should be if check_price == True

File: src/encoded/commands/specification copy.json [hunk 2]
```diff
@@ -0,0 +1,13 @@
+{
```
Comment (hitz) [inline src/encoded/commands/specification copy.json]: delete this file

File: src/encoded/commands/deploy.py [hunk 3]
```diff
@@ -117,26 +293,48 @@ def run(wale_s3_prefix, image_id, instance_type, elasticsearch, cluster_size, cl
...
 
+        if spot_instance == False :    
+            print('%s.%s.encodedcc.org' % (instance.id, domain))  # Instance:i-34edd56f
+            instance.wait_until_exists()
+            tag_ec2_instance(instance, tmp_name, branch, commit, username, elasticsearch, cluster_name)
+            print('ssh %s.%s.encodedcc.org' % (tmp_name, domain))
+            if domain == 'instance':
+                print('https://%s.demo.encodedcc.org' % tmp_name)
 
+    if not spot_instance == False:
```
Comment (hitz) [inline src/encoded/commands/deploy.py]: if spot_instance == True... or maybe an elif on line 327?

File: src/encoded/commands/deploy.py [hunk 4]
```diff
@@ -117,26 +293,48 @@ def run(wale_s3_prefix, image_id, instance_type, elasticsearch, cluster_size, cl
...
         iam_role = 'elasticsearch-instance'
         count = int(cluster_size)
+    
+    if not check_price == False :
+        ec2_spot = boto3.client('ec2')
+        get_spot_price = spot_instance_price_check(ec2_spot, instance_type)
+        exit()
 
-    instances = create_ec2_instances(ec2, image_id, count, instance_type, security_groups, user_data, BDM, iam_role)
+    if not spot_instance == False :
```
Comment (hitz) [inline src/encoded/commands/deploy.py]: if spot_instance == True

File: src/encoded/commands/deploy.py [hunk 5]
```diff
@@ -117,26 +293,48 @@ def run(wale_s3_prefix, image_id, instance_type, elasticsearch, cluster_size, cl
...
+        print("spot_instance check worked")
+        spot_security_groups = 'ssh-http-https'
```
Comment (hitz) [inline src/encoded/commands/deploy.py]: Why is this different from security_groups set above?  The have the same value?

File: src/encoded/commands/deploy.py [hunk 6]
```diff
@@ -54,11 +215,26 @@ def tag_ec2_instance(instance, name, branch, commit, username, elasticsearch):
     ]
     if elasticsearch == 'yes':
         tags.append({'Key': 'elasticsearch', 'Value': elasticsearch})
+    if not cluster_name == None:
```
Comment (hitz) [inline src/encoded/commands/deploy.py]: if cluster_name != None

File: src/encoded/commands/deploy.py [hunk 7]
```diff
@@ -23,7 +27,164 @@
...
+        if key == 'SpotInstanceRequests':
+            for item in value:
+                for i in item:
+                    if i == 'SpotInstanceRequestId':
+                        SpotInstanceRequestId=item[i]
+    return SpotInstanceRequestId
+
+def get_spot_code(instance, client, spot_id):
+    request = client.describe_spot_instance_requests(SpotInstanceRequestIds=[get_spot_id(instance, client)])
+    for key, value in request.items():
```
Comment (hitz) [inline src/encoded/commands/deploy.py]: There is a way cleaner way to do this loop.
I think what you are looking for is:
code_status = request['SpotInstnceRequests]['Status']['Code']
I don't think you need to iterate.

File: src/encoded/commands/deploy.py [hunk 8]
```diff
@@ -23,7 +27,164 @@
...
+    def spotClient(self):
+        return self._spotClient
+
+    @spotClient.setter
+    def spotClient(self, value):
+        self._spotClient = value
+
+def get_spot_id(instance, client):
+    
+    for key, value in instance.items():
```
Comment (hitz) [inline src/encoded/commands/deploy.py]: this is just instance['SpotInstanceRequests']['SpotInstanceRequestId']

File: src/encoded/commands/deploy.py [hunk 9]
```diff
@@ -3,6 +3,10 @@
 import re
 import subprocess
 import sys
+import datetime
+import base64
+import pdb
```
Comment (hitz) [inline src/encoded/commands/deploy.py]: remove

File: src/encoded/commands/deploy.py [hunk 10]
```diff
@@ -3,6 +3,10 @@
...
+from pprint import pprint as pp
```
Comment (hitz) [inline src/encoded/commands/deploy.py]: remove

File: src/encoded/commands/deploy copy.py [hunk 11]
```diff
@@ -0,0 +1,183 @@
+import boto3
```
Comment (hitz) [inline src/encoded/commands/deploy copy.py]: Delete this file, don't check in copies.  Git has all history of all changes, trust it!!!

File: scripts/stress.py [hunk 12]
```diff
@@ -0,0 +1,38 @@
+from multiprocessing import Pool
```
Comment (hitz) [inline scripts/stress.py]: I think you moved this in another branch, probably want to merge that in.

File: scripts/locustfile.py [hunk 13]
```diff
@@ -0,0 +1,50 @@
+from locust import HttpLocust, TaskSet, task
```
Comment (hitz) [inline scripts/locustfile.py]: I think you moved this in another branch...

File: src/encoded/commands/deploy copy.py (hunk 11 above)
Comment (gto389) [inline src/encoded/commands/deploy copy.py]: done

File: src/encoded/commands/specification copy.json (hunk 2 above)
Comment (gto389) [inline src/encoded/commands/specification copy.json]: done

File: src/encoded/commands/deploy.py (hunk 3 above)
Comment (gto389) [inline src/encoded/commands/deploy.py]: done

File: src/encoded/commands/deploy.py (hunk 5 above)
Comment (gto389) [inline src/encoded/commands/deploy.py]: done

File: src/encoded/commands/deploy.py (hunk 1 above)
Comment (gto389) [inline src/encoded/commands/deploy.py]: done

File: src/encoded/commands/deploy.py (hunk 6 above)
Comment (gto389) [inline src/encoded/commands/deploy.py]: done

File: src/encoded/commands/deploy.py (hunk 10 above)
Comment (gto389) [inline src/encoded/commands/deploy.py]: Need base64 so the encoding can be set and returned correctly. Date and time is used for checking todays prices. The others are gone.

File: src/encoded/commands/deploy.py (hunk 9 above)
Comment (gto389) [inline src/encoded/commands/deploy.py]: done

File: scripts/locustfile.py (hunk 13 above)
Comment (hitz) [inline scripts/locustfile.py]: never mind, I am confused by other branch.

File: scripts/stress.py (hunk 12 above)
Comment (hitz) [inline scripts/stress.py]: see previous comment

**Pull request merged**